dimagi/commcare-hq
# Riese/concat session endpoint

## Product Description

Add the ability to set session endpoint ids for forms in shadow modules.

<img width="800" alt="image" src="https://github.com/dimagi/commcare-hq/assets/1946138/c9f99e4e-6fc7-4be3-b798-1ff329c94ac8">

## Technical Summary

The session endpoint ids for forms in shadow modules are stored in `form_session_endpoint_mapping` which mapsform id to session endpoint id. At first I had planned to combine this with `excluded_form_ids`. But this would change the behavior when adding new forms to the source module (see spec document).

Because in the case of shadow modules now more than one session endpoint id can be set in a single request, I added `set_shadow_module_and_form_session_endpoint` to `utils.py` that handles the validation for this case.

Spec:
https://docs.google.com/document/d/1hnOlnvEa0EHlsTMgqZsiOU8Te96t2bJlVKtYqf1jwhw/edit

Jira Story:
https://dimagi-dev.atlassian.net/browse/USH-3395

## Feature Flag

* Shadow Modules
* Enable session endpoints

## Safety Assurance

### Safety story

Tested locally and in staging.

### Automated test coverage

Added a new test that covers the refactored and new functions in `utils.py`
corehq/apps/app_manager/tests/views/test_utils.py

### QA Plan

1. Create app with a main menu, sub/child menu, shadow menu of main menu, and shadow menu of sub/child menu.
2. Add New Entry and Follow up forms to main module and at least a Follow up form to the child module.
3. Set a session endpoint id on all modules and forms. Reusing any of the other session endpoint ids should result in a validation error.
4. Publish the app and create a smart link using all of the session endpoint ids and make sure that they are working.

Tested using this [app in staging](https://staging.commcarehq.org/a/mriese/apps/view/a671db914232020002c1f294c651481d/)

**Main menu**
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/main-menu => main menu
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/main-registration => registration form
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/main-follow/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00  =>  follow up form on main menu with case preselected

**Sub/Child menu**
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/child-menu/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 => child menu with case preselected
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/child-follow/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 => follow up form on child menu with case preselected

**Shadow menu of main menu**
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow =>  shadow menu
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-registration => registration form under shadow menu
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-follow/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 => follow up form under shadow menu with case preselected

**Sub/Child menu of shadow menu**
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-child/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 =>  child menu of shadow menu with case preselected
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-child-follow/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 =>  follow up form on child menu under shadow menu with case preselected

**Shadow menu of sub/child menu**
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-of-child/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 =>  shadow of child menu with case preselected
https://staging.commcarehq.org/a/mriese/app/v1/a671db914232020002c1f294c651481d/shadow-of-child-follow/?case_id=49756a4a-4ea3-411c-b942-2d6fbf19ad00 => follow up form of shadow of child menu with case preselected

### Rollback instructions




### Labels & Review
